### ISSUE TITLE
PasswordReset verification finds finds wrong row

### ISSUE THREAD

**Issue by fleeser** [serverpod/serverpod#643]
When verifying the PasswordReset for email authentication, the following happens:

1. Find the row where `verificationCode == inputVerificationCode`
2. Get the `UserInfo` based on the `userId` from that row
3. Send back the found UserInfo

```
static Future<EmailPasswordReset?> verifyEmailPasswordReset(
  Session session,
  String verificationCode,
) async {
  session.log('verificationCode: $verificationCode', level: LogLevel.debug);
  var passwordReset = await session.db.findSingleRow<EmailReset>(
    where: EmailReset.t.verificationCode.equals(verificationCode) &
        (EmailReset.t.expiration > DateTime.now().toUtc()),
  );

  if (passwordReset == null) return null;

  var userInfo = await Users.findUserByUserId(session, passwordReset.userId);
  if (userInfo == null) return null;

  if (userInfo.email == null) return null;

  return EmailPasswordReset(
    userName: userInfo.userName,
    email: userInfo.email!,
  );
}
```

Since there is no check on if this is the `verificationCode` of the requesting user, I see the following two problems:

- When entering a `verificationCode` randomly there is a chance of verifying someone else's `PasswordReset`, because the `verificationCode` is generated randomly (even if chances are low).
- The user would get back wrong `UserInfo` and later in that process change someone else's password.

**Solution**

I think one solution could be to pass the `email` to the endpoint which the user's device knows from when they requested the reset instead of the `userId` and store the `email` inside the `EmailReset` table. In the `verifyEmailPasswordReset` then, we should find the row where the `email` is equal to the one that is passed. We could then find the `UserInfo` based on the `email` instead of the `userId` and pass it back.

**Serverpod versions**
 -  _CLI Version:_ 1.0.0
 - _Library version:_ 1.0.0

**Comment by tp** [comment]
Looked into this while working on "Auth 2.0", to capture the current state:


The verification code is unique, so the server would never use the same code for 2 users:

https://github.com/serverpod/serverpod/blob/d13ae968a342541293dcde58f836c24635a10481/modules/serverpod_auth/serverpod_auth_server/lib/src/models/email_reset.spy.yaml#L14-L16

As for the length and randomly guessing it: Currently, this is controlled by `validationCodeLength`, which has a default length of 8 characters, and a required minimum length of 4:

https://github.com/serverpod/serverpod/blob/d13ae968a342541293dcde58f836c24635a10481/modules/serverpod_auth/serverpod_auth_server/lib/src/business/email_auth.dart#L402-L405

Additionally, the verification code is only valid for 15 minutes (by default), though there is no upper limit on what could be configured:

https://github.com/serverpod/serverpod/blob/d13ae968a342541293dcde58f836c24635a10481/modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart#L198

---

If we assume the user stays "in flow" (e.g. inputs the code with the same session that they just used for the request), we could of course send something along (e.g. the request ID or the email again etc.). But I think we have to be careful anyway to not allow shorter codes, so that this can not be used by an attacker to guess the code after they requested a "fake" reset (in which case they would know the email (or password reset ID) they want to use already).

**Comment by Isakdl** [comment]
@tp Good points, but to add to that if we did have the email or request id, we could limit the number of tries with the code. Lets say 3 and then you would have to make a new password reset request. On top of that we could limit the number of password reset request within a certain time window let say 1h, 24h etc. 

This way it is impossible to brute force given a sufficient length of the code itself.

**Comment by tp** [comment]
@Isakdl Right, those additional safeguards should definitely be layered on top. I'll see what we have now and ensure to match/improve that for the new module.

I think the last part is really about the UX / user flow. Is the code something we expect the user to read in the e-mail app, and then copy/paste or potentially type (into a second device) to reset the password with, or would we use a code that is only intended to be used in links (either web or deep link to app) where the user then gets to set the new password (for a few minutes).

I think I mostly see the latter approach in the wild, and there is then really no limit to how long we could make it, if the user never has to transfer it manually.

**Comment by Isakdl** [comment]
Agreed I think the latter is most likely what most people want, I would personally prefer building like that.

**Comment by marcelomendoncasoares** [comment]
Closing this since the issue is solved in the new auth.
